README.rst
collective.autoqueryplan
====================

.. image:: https://secure.travis-ci.org/collective/collective.autoqueryplan.png
   :target: http://travis-ci.org/collective/collective.autoqueryplan

*collective.autoqueryplan* dumps a json version of all the zcatalogs queryplans
to a file on disk periodically. If this file is found during load then it is
used to prime the catalogs query plans instead of the queryplan specified in
```ZCATALOGQUERYPLAN``` or having an empty queryplan (zopes default).

The advantage of a queryplan is it can help speedup catalog searches and reduce
unneeded ZODB loads by knowing more about most common queries an instance is
making. ```ZCATALOGQUERYPLAN``` allows for a single snapshot applied across
all instances and has to be manually updated. Autoqueryplan allows for
different instances to have different queryplans overtime and doesn't
require any intervention. For example p.a.async worker instances can have
very different query patterns.

Minimal configuration:

.. code:: ini

    zope-conf-additional =
            <clock-server>
               method /@@dumpqueryplan
               period 3600
            </clock-server>


Advanced configuration
----------------------

By default the file is stored in ${client_home}/queryplan.json. You can override this with the environment variable:

``AUTOQUERYPLAN`` *(default=${client_home}/queryplan.json)*
    Where to dump and load queryplan json from


CHANGES.txt
Changelog
=========


0.1.0 (2019-05-25)
------------------

- First release for experimental use.


src/collective/autoqueryplan/tests/test_pmap.py
# -*- coding: utf-8 -*-
import doctest





def test_suite():
    suite = doctest.DocTestSuite('collective.autoqueryplan.server')
    return suite


src/collective/autoqueryplan/tests/test_task_id.py
# -*- coding: utf-8 -*-
import Queue
import logging
import unittest

from plone.testing import z2
import transaction
from zope.component import getUtility
from zope.testing.loggingsupport import InstalledHandler

from collective.autoqueryplan import autoqueryplan
from collective.autoqueryplan.interfaces import Iautoqueryplan
from collective.autoqueryplan.testing import (
    LocalautoqueryplanServerLayer,
    runAsyncTest
)


logger = logging.getLogger('collective.autoqueryplan')


class TaskIdLoggingautoqueryplanServerLayer(LocalautoqueryplanServerLayer):

    def setUp(self):
        super(TaskIdLoggingautoqueryplanServerLayer, self).setUp()

        def logging_handler(app, request, response):
            logger.info(request.getHeader("X-Task-Id"))
            response.stdout.write('HTTP/1.1 204\r\n')
            response.stdout.close()

        self['server'].handler = logging_handler


TASK_QUEUE_FIXTURE = TaskIdLoggingautoqueryplanServerLayer(queue='test-queue')

TASK_QUEUE_FUNCTIONAL_TESTING = z2.FunctionalTesting(
    bases=(TASK_QUEUE_FIXTURE,),
    name='autoqueryplan:Functional')


class TestLocalVolatileautoqueryplan(unittest.TestCase):

    layer = TASK_QUEUE_FUNCTIONAL_TESTING
    queue = 'test-queue'

    @property
    def task_queue(self):
        return getUtility(Iautoqueryplan, name=self.queue)

    def setUp(self):
        self.task_queue.queue = Queue.Queue()

    def _testConsumeFromQueue(self):
        self.assertEqual(len(self.task_queue), 0)

    def testTaskId(self):
        self.assertEqual(len(self.task_queue), 0)
        a = autoqueryplan.add("/", queue=self.queue)
        b = autoqueryplan.add("/Plone", queue=self.queue)
        transaction.commit()
        self.assertEqual(len(self.task_queue), 2)

        handler = InstalledHandler("collective.autoqueryplan")
        runAsyncTest(self._testConsumeFromQueue)
        messages = [record.getMessage() for record in handler.records]
        self.assertEqual(messages[-2:], [a, b])



src/collective/autoqueryplan/tests/test_taskqueue.py
# -*- coding: utf-8 -*-
import Queue
import transaction
import unittest2 as unittest
from zope.component import getUtility
from zope.testing.loggingsupport import InstalledHandler

from collective.autoqueryplan.interfaces import Iautoqueryplan
from collective.autoqueryplan.testing import runAsyncTest
from collective.autoqueryplan.testing import TASK_QUEUE_FUNCTIONAL_TESTING
from collective.autoqueryplan import autoqueryplan


class TestLocalVolatileautoqueryplan(unittest.TestCase):

    layer = TASK_QUEUE_FUNCTIONAL_TESTING
    queue = 'test-queue'

    @property
    def task_queue(self):
        return getUtility(Iautoqueryplan, name=self.queue)

    def setUp(self):
        self.task_queue.queue = Queue.Queue()

    def testEmptyQueue(self):
        self.assertEqual(len(self.task_queue), 0)

    def testAddToQueue(self):
        autoqueryplan.add("/", queue=self.queue)
        self.assertEqual(len(self.task_queue), 0)

    def testCommitToQueue(self):
        autoqueryplan.add("/", queue=self.queue)
        self.assertEqual(len(self.task_queue), 0)
        transaction.commit()
        self.assertEqual(len(self.task_queue), 1)
        autoqueryplan.add("/Plone", queue=self.queue)
        self.assertEqual(len(self.task_queue), 1)
        transaction.commit()
        self.assertEqual(len(self.task_queue), 2)

    def _testConsumeFromQueue(self):
        self.assertEqual(len(self.task_queue), 0)

    def testConsumeFromQueue(self):
        self.assertEqual(len(self.task_queue), 0)
        autoqueryplan.add("/", queue=self.queue)
        autoqueryplan.add("/Plone", queue=self.queue)
        transaction.commit()
        self.assertEqual(len(self.task_queue), 2)

        handler = InstalledHandler("collective.autoqueryplan")
        runAsyncTest(self._testConsumeFromQueue)
        messages = [record.getMessage() for record in handler.records]
        self.assertEqual(messages[-2:],
                         ["http://nohost:/", "http://nohost:/Plone"])

    def testConsume100FromQueue(self):
        self.assertEqual(len(self.task_queue), 0)
        expected_result = []
        for i in range(100):
            autoqueryplan.add("/{0:02d}".format(i), queue=self.queue)
            expected_result.append("http://nohost:/{0:02d}".format(i))
        transaction.commit()
        self.assertEqual(len(self.task_queue), 100)

        handler = InstalledHandler("collective.autoqueryplan")
        runAsyncTest(self._testConsumeFromQueue)
        messages = [record.getMessage() for record in handler.records]

        self.assertEqual(sorted(messages[-100:]), expected_result)


src/collective/autoqueryplan/server.py
# -*- coding: utf-8 -*-
import ast
import asyncore
import json
import os
import time
import socket
import StringIO
import posixpath
import logging
import threading

from App.config import getConfiguration
from Products.ZCatalog.plan import PriorityMap, Benchmark
from ZServer.ClockServer import LogHelper, ClockServer, timeslice
from ZServer.PubCore.ZEvent import Wakeup
from ZServer.medusa.http_server import http_request
from ZServer.medusa.default_handler import unquote
from ZServer.PubCore import handle
from ZServer.HTTPResponse import make_response
from ZPublisher.HTTPRequest import HTTPRequest
from zope.component import getUtility
from zope.component import ComponentLookupError
from zope.dottedname.resolve import resolve
from zope.interface import implements
from zope.publisher.browser import BrowserView

logger = logging.getLogger('collective.autoqueryplan')


def get_plan_path():
    path = os.environ.get('AUTOQUERYPLAN', None)
    if path is None:
        path = getConfiguration().clienthome
        path = os.path.join(path, 'queryplan.json')
    return path


def dump_pmap(obj, is_key=False):
    """
    Can't have sets, classes or complex objects as dict keys in json.

    Handle complex keys
    >>> dump_pmap({frozenset([1,2,3]):"blah"})
    {'{1, 2, 3}': 'blah'}

    >>> dump_pmap({(1,2,3):set(['a'])})
    {'(1, 2, 3)': "{'a'}"}

    Handles tuples as keys (which json can't)
    #TODO: it always assumes sets are frozen sets
    >>> load_pmap(dump_pmap({(1,2,3):set(['a'])}))
    {(1, 2, 3): frozenset(['a'])}

    Works for lists of sets in (even if not needed for this usecase
    >>> dump_pmap([1,2,set(['a'])])
    [1, 2, "{'a'}"]

    >>> load_pmap(dump_pmap([1,2,set(['a'])]))
    [1, 2, frozenset(['a'])]

    """
    if isinstance(obj, Benchmark):
        return (round(obj[0], 4),) + obj[1:]
    elif isinstance(obj, (frozenset, set)):
        return "{%s}" % str(list(obj))[1:-1]
    elif is_key and isinstance(obj, (list, tuple)):
        return str(obj)
    elif isinstance(obj, (tuple)):
        return (dump_pmap(i) for i in obj)
    elif isinstance(obj, (list)):
        return [dump_pmap(i) for i in obj]
    elif isinstance(obj, dict):
        return {dump_pmap(key, True): dump_pmap(value) for key, value in obj.items()}
    else:
        return obj

def load_pmap(obj):
    """ turn complex dict keys from strings back to types """
    if isinstance(obj, basestring):
        if obj.startswith('{'):
            return frozenset(ast.literal_eval("[%s]" % obj[1:-1]))
        elif obj.startswith('frozenset(['):
            return frozenset(ast.literal_eval(obj[len('frozenset('):-1]))
        elif obj.startswith('(') or obj.startswith('['):
            return ast.literal_eval(obj)
        else:
            return obj
    elif isinstance(obj, (tuple)):
        return (load_pmap(i) for i in obj)
    elif isinstance(obj, (list)):
        return [load_pmap(i) for i in obj]
    elif isinstance(obj, dict):
        return {load_pmap(key): load_pmap(value) for key, value in obj.items()}
    else:
        return obj


_load_default = PriorityMap.load_default

def load_dumped_plan(cls):
    path = get_plan_path()
    plan = None
    try:
        with open(path) as f:
            plan = f.read()
    except:
        logger.error("Problem reading queryplan %s" % path)
    if plan is not None:
        try:
            pmap = json.loads(plan)
            pmap = load_pmap(pmap)
            cls.load_pmap(path, pmap)
            return
        except Exception:
            #TODO: should put more sepecifc exceptions
            logger.error("Problem loading queryplan %s" % path)
    return _load_default(cls)

# Monkey patch querymap loading so we load autoqueryplan first
PriorityMap.load_default = classmethod(load_dumped_plan)


class DumpPlanView(BrowserView):
    def __call__(self):
        pmap = PriorityMap.get_value()
        pmap = dump_pmap(pmap)
        json_pmap = json.dumps(pmap, skipkeys=False, sort_keys=True, indent=4)
        with open(get_plan_path(),'w') as f:
            f.write(json_pmap)
        logger.info("Succesfully dumped queryplan to %s" % get_plan_path())
        return "Query plan dumped"

#
# class QueryPlanServer(ClockServer):
#     """ every hour dump a json version of the queryplan to disk """
#
#     pass
#
# class ClockServer(asyncore.dispatcher):
#     # prototype request environment
#     _ENV = dict(REQUEST_METHOD = 'GET',
#                 SERVER_PORT = 'Clock',
#                 SERVER_NAME = 'Zope Clock Server',
#                 SERVER_SOFTWARE = 'Zope',
#                 SERVER_PROTOCOL = 'HTTP/1.0',
#                 SCRIPT_NAME = '',
#                 GATEWAY_INTERFACE='CGI/1.1',
#                 REMOTE_ADDR = '0')
#
#     # required by ZServer
#     SERVER_IDENT = 'Zope Clock'
#
#     def __init__ (self, period=60, logger=None, handler=None):
#         self.period = period
#
#         self.last_slice = timeslice(period)
#
#
#         asyncore.dispatcher.__init__(self)
#         self.create_socket(socket.AF_INET, socket.SOCK_STREAM)
#         self.logger = LogHelper(logger)
#         self.log_info('Clock server for auoqueryplan started (period: %s)'
#                       % (self.period))
#         if handler is None:
#             # for unit testing
#             handler = handle
#         self.zhandler = handler
#
#
#     def readable(self):
#         # generate a request at most once every self.period seconds
#         slice = timeslice(self.period)
#         if slice != self.last_slice:
#             # no need for threadsafety here, as we're only ever in one thread
#             self.last_slice = slice
#             req, zreq, resp = self.get_requests_and_response()
#             self.zhandler('Zope2', zreq, resp)
#         return False
#
#     def handle_read(self):
#         return True
#
#     def handle_write (self):
#         self.log_info('unexpected write event', 'warning')
#         return True
#
#     def writable(self):
#         return False
#
#     def handle_error (self):      # don't close the socket on error
#         (file,fun,line), t, v, tbinfo = asyncore.compact_traceback()
#         self.log_info('Problem in Clock (%s:%s %s)' % (t, v, tbinfo),
#                       'error')
#


src/collective/autoqueryplan/testing.py
# -*- coding: utf-8 -*-
import asyncore
import logging
from App.config import getConfiguration

from zope.configuration import xmlconfig
from zope.component import getSiteManager

from plone.testing import Layer
from plone.testing import z2
from collective.autoqueryplan.config import HAS_REDIS
from collective.autoqueryplan.config import HAS_MSGPACK
from collective.autoqueryplan import autoqueryplan
from collective.autoqueryplan.interfaces import Iautoqueryplan

if HAS_REDIS and HAS_MSGPACK:
    from collective.autoqueryplan import redisqueue

logger = logging.getLogger('collective.autoqueryplan')

class autoqueryplanServerLayer(Layer):
    defaultBases = (z2.STARTUP,)

    def __init__(self, queue='test-queue', zserver_enabled=False):
        super(autoqueryplanServerLayer, self).__init__()
        self.queue = queue
        self.zserver_enabled = zserver_enabled

    def setUp(self):
        import collective.autoqueryplan
        xmlconfig.file('configure.zcml', collective.autoqueryplan,
                       context=self['configurationContext'])

        # Configure
        config = getConfiguration()
        config.product_config = {'collective.autoqueryplan': {'queue': self.queue}}
        autoqueryplan.reset()

        # Define logging request handler to replace ZPublisher
        def logging_handler(app, request, response):
            logger.info(request.getURL() + request.get("PATH_INFO"))
            response.stdout.write('HTTP/1.1 204\r\n')
            response.stdout.close()

        # Define ZPublisher-based request handler to be used with zserver
        def zserver_handler(app, request, response):
            from ZPublisher import publish_module
            publish_module(app, request=request, response=response)

        # Create autoqueryplanServer
        from collective.autoqueryplan.server import autoqueryplanServer
        if not self.zserver_enabled:
            self['server'] = autoqueryplanServer(queue=self.queue,
                                             handler=logging_handler)
        else:
            self['server'] = autoqueryplanServer(queue=self.queue,
                                             handler=zserver_handler,
                                             concurrent_limit=0)
            # concurrent_limit=0, because of limitations in z2.ZServer

    def tearDown(self):
        self['server'].handle_close(force=True)

    def testTearDown(self):
        autoqueryplan.reset()


class LocalautoqueryplanServerLayer(autoqueryplanServerLayer):

    def setUp(self):
        queue = autoqueryplan.LocalVolatileautoqueryplan()
        sm = getSiteManager()
        sm.registerUtility(queue, provided=Iautoqueryplan, name='test-queue')
        super(LocalautoqueryplanServerLayer, self).setUp()


TASK_QUEUE_FIXTURE = LocalautoqueryplanServerLayer(queue='test-queue')
TASK_QUEUE_ZSERVER_FIXTURE =\
    LocalautoqueryplanServerLayer(queue='test-queue', zserver_enabled=True)

TASK_QUEUE_INTEGRATION_TESTING = z2.IntegrationTesting(
    bases=(TASK_QUEUE_FIXTURE,),
    name='autoqueryplan:Integration')

TASK_QUEUE_FUNCTIONAL_TESTING = z2.FunctionalTesting(
    bases=(TASK_QUEUE_FIXTURE,),
    name='autoqueryplan:Functional')


class RedisautoqueryplanServerLayer(autoqueryplanServerLayer):

    def setUp(self):
        queue = redisqueue.Redisautoqueryplan()
        sm = getSiteManager()
        sm.registerUtility(queue, provided=Iautoqueryplan, name='test-queue')
        super(RedisautoqueryplanServerLayer, self).setUp()

REDIS_TASK_QUEUE_FIXTURE = RedisautoqueryplanServerLayer(queue='test-queue')
REDIS_TASK_QUEUE_ZSERVER_FIXTURE =\
    RedisautoqueryplanServerLayer(queue='test-queue', zserver_enabled=True)

REDIS_TASK_QUEUE_INTEGRATION_TESTING = z2.IntegrationTesting(
    bases=(REDIS_TASK_QUEUE_FIXTURE,),
    name='Redisautoqueryplan:Integration')

REDIS_TASK_QUEUE_FUNCTIONAL_TESTING = z2.FunctionalTesting(
    bases=(REDIS_TASK_QUEUE_FIXTURE,),
    name='Redisautoqueryplan:Functional')
